pkg/iam/iam.go

```diff
@@ -0,0 +1,399 @@
+		var a map[string]interface{}
+		err = json.Unmarshal(userPolicy, &a)
+		for _, statement := range a["Statement"].([]interface{}) {
+			elem := statement.(map[string]interface{})
+			if elem["Effect"] == "Allow" {
+				var singleAction bool
+				var allResources bool
+				_, ok := elem["Resource"].(string)
+				if ok {
+					if elem["Resource"] == "*" {
+						allResources = true
+					}
+				}
+				_, ok = elem["Action"].(string)
+				if ok {
+					singleAction = true
+				}
+				if allResources && singleAction {
+					singleact, ok := elem["Action"].(string)
+					if ok {
+						permissions = append(permissions, singleact)
+					}
+				} else if allResources {
+					for _, action := range elem["Action"].([]interface{}) {
+						act, ok := action.(string)
+						if ok {
+							permissions = append(permissions, act)
+						} else {
+							logrus.Fatalf("Error appending action %q to permissions list: %v", action, permissions)
+						}
+					}
+				}
+			}
+		}
+	}
+	return permissions
+}
+
+// CheckOffPermissionsByPrefix accepts a map of permissions
+// to be used as a checklist. It then marks all permissions 'true'
+// if they match the specified prefix.
+func CheckOffPermissionsByPrefix(checklist map[string]bool, prefix string) map[string]bool {
+	if prefix == "*" {
+		for k := range checklist {
+			checklist[k] = true
+		}
+	}
+	for k := range checklist {
+		match, _ := regexp.MatchString(prefix, k)
+		if match {
+			logrus.Debugf("Prefix %q matches checklist item %q", prefix, k)
+			checklist[k] = true
+		}
+	}
+	return checklist
+}
+
+// RunPermissionsCheck ties together all the above helper functions
+// to run a complete suite of AWS IAM permissions checks. This will
+// inspect permissions attached directly to users and groups, as well
+// as managed policies containing permissions granted to the user.
+func RunPermissionsCheck() {
+
+	// Create an IAM service client.
+	sess, err := session.NewSession(&aws.Config{})
+	if err != nil {
+		logrus.Fatalf("Error creating new AWS session: %s", err)
+	}
+	svc := iam.New(sess)
+
+	accountAuthDetails := GetAccountAuthDetails(svc)
+	user := GetCurrentUserDetails(svc, accountAuthDetails)
+	logrus.Infof("Current IAM user: %s", *user.UserName)
+	logrus.Debugf("User details:\n %s", user)
+
+	// checklist is a map used as a checklist, containing all
+	// permissions required to run both the installer and uninstaller.
+	checklist := make(map[string]bool)
+	for _, p := range Actions {
+		checklist[p] = false
+	}
+
+	// userPolicyDocs is a slice that stores the Policy Documents
+	// (permissions details) associated with the current user.
+	var userPolicyDocs []string
+
+	GetPolicyDocFromUser(svc, user, &userPolicyDocs, accountAuthDetails)
+	GetPolicyDocFromGroup(svc, user, &userPolicyDocs, accountAuthDetails)
+
+	userPermissions := GetPermissionsFromPolicyDoc(&userPolicyDocs)
+
+	for _, permission := range userPermissions {
+		CheckOffPermissionsByPrefix(checklist, permission)
+	}
+
+	var missing bool
```

Comment:
nit: unset `bool`s do default to `false` in golang. For those who are newer to the language who may read through I believe it's nice to write the statement as `missing := false` as is easier to understand without knowing golang defaults. Doesn't need to be changed but thought I'd make a note :smile:.